README
In this directory:

html-xml-utils-*.tar.gz
    A number of simple utilities for manipulating HTML and XML files.
    See INSTALL for generic installation instructions.
    Get the source at: http://www.w3.org/Tools/HTML-XML-utils/

htmlutils-*.tar.gz
    Old versions (before version 0.1)


Note 1: Your package manager may have a precompiled copy
already. There are versions in Debian, Ubuntu, Macports and others. In
that case you need to download from here only if you want a different
version or want to hack on the source.

Note 2: the names changed in version 5.0: most programs got an "hx"
prefix. Please, uninstall any version < 5.0 before installing a
version >= 5.0


cexport (1)          - create headerfile of exported declarations from a C file
hxaddid (1)          - add ID's to selected elements
hxcite (1)           - replace bibliographic references by hyperlinks
hxcite-mkbib (1)     - expand references and create bibliography
hxcopy (1)           - copy an HTML file while preserving relative links
hxcount (1)          - count elements and attributes in HTML or XML files
hxextract (1)        - extract selected elements
hxclean (1)          - apply heuristics to correct an HTML file
hxprune (1)          - remove marked elements from an HTML file
hxincl (1)           - expand included HTML or XML files
hxindex (1)          - create an alphabetically sorted index
hxmkbib (1)          - create bibliography from a template
hxmultitoc (1)       - create a table of contents for a set of HTML files
hxname2id            - move some ID= or NAME= from A elements to their parents
hxnormalize (1)      - pretty-print an HTML file
hxnum (1)            - number section headings in an HTML file
hxpipe (1)           - convert XML to a format easier to parse with Perl or AWK
hxprintlinks (1)     - number links & add table of URLs at end of an HTML file
hxremove (1)         - remove selected elements from an XML file
hxtabletrans (1)     - transpose an HTML or XHTML table
hxtoc (1)            - insert a table of contents in an HTML file
hxuncdata (1)        - replace CDATA sections by character entities
hxunent (1)          - replace HTML predefined character entities to UTF-8
hxunpipe (1)         - convert output of pipe back to XML format
hxunxmlns (1)        - replace "global names" by XML Namespace prefixes
hxwls (1)            - list links in an HTML file
hxxmlns (1)          - replace XML Namespace prefixes by "global names"
asc2xml, xml2asc (1) - convert between UTF8 and &#nnn; entities
hxref (1)            - generate cross-references
hxselect (1)         - extract elements that match a (CSS) selector



This package is configured with automake/autoconf. Generic
instructions are in the file INSTALL. Here are some specific problems
that may arise:

1) Error when running lex:

      lex   scan.l && mv lex.yy.c scan.c
      "scan.l":line 2: Error: missing translation value

   The scan.l file uses features of flex that do not exist in lex.
   However, it is not necessary to run lex, since the file scan.c is
   provided in the package. Just do a "touch scan.c" to make sure
   "make" will not try to generate it anew.

2) Warning about "libidn not found":

  Without libidn2 or libidn, hxwls will not be able to translate
  Internationalized Domain Names to ASCII (option -a). You can install
  either libidn2 or libidn.

  If you install them in a non-standard location, use --with-libidn2
  or --with-libidn when invoking ./configure. E.g., if you install
  libidn from MacPorts on Mac OS X, run:

      ./configure --with-libidn=/opt/local


$Date: 2016/04/14 00:42:15 $


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash

dtd.c
/* C code produced by gperf version 3.0.4 */
/* Command-line: gperf -a -c -C -o -t -p -T -k '1,2,$' -N lookup_element dtd.hash  */

#if !((' ' == 32) && ('!' == 33) && ('"' == 34) && ('#' == 35) \
      && ('%' == 37) && ('&' == 38) && ('\'' == 39) && ('(' == 40) \
      && (')' == 41) && ('*' == 42) && ('+' == 43) && (',' == 44) \
      && ('-' == 45) && ('.' == 46) && ('/' == 47) && ('0' == 48) \
      && ('1' == 49) && ('2' == 50) && ('3' == 51) && ('4' == 52) \
      && ('5' == 53) && ('6' == 54) && ('7' == 55) && ('8' == 56) \
      && ('9' == 57) && (':' == 58) && (';' == 59) && ('<' == 60) \
      && ('=' == 61) && ('>' == 62) && ('?' == 63) && ('A' == 65) \
      && ('B' == 66) && ('C' == 67) && ('D' == 68) && ('E' == 69) \
      && ('F' == 70) && ('G' == 71) && ('H' == 72) && ('I' == 73) \
      && ('J' == 74) && ('K' == 75) && ('L' == 76) && ('M' == 77) \
      && ('N' == 78) && ('O' == 79) && ('P' == 80) && ('Q' == 81) \
      && ('R' == 82) && ('S' == 83) && ('T' == 84) && ('U' == 85) \
      && ('V' == 86) && ('W' == 87) && ('X' == 88) && ('Y' == 89) \
      && ('Z' == 90) && ('[' == 91) && ('\\' == 92) && (']' == 93) \
      && ('^' == 94) && ('_' == 95) && ('a' == 97) && ('b' == 98) \
      && ('c' == 99) && ('d' == 100) && ('e' == 101) && ('f' == 102) \
      && ('g' == 103) && ('h' == 104) && ('i' == 105) && ('j' == 106) \
      && ('k' == 107) && ('l' == 108) && ('m' == 109) && ('n' == 110) \
      && ('o' == 111) && ('p' == 112) && ('q' == 113) && ('r' == 114) \
      && ('s' == 115) && ('t' == 116) && ('u' == 117) && ('v' == 118) \
      && ('w' == 119) && ('x' == 120) && ('y' == 121) && ('z' == 122) \
      && ('{' == 123) && ('|' == 124) && ('}' == 125) && ('~' == 126))
/* The character set is not based on ISO-646.  */
error "gperf generated tables don't work with this execution character set. Please report a bug to <bug-gnu-gperf@gnu.org>."
#endif

#line 1 "dtd.hash"
						/* -*-indented-text-*- */

/*
 * Copyright © 1998-2017 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/copyright-software
 *
 * Part of HTML-XML-utils, see:
 * http://www.w3.org/Tools/HTML-XML-utils/
 *
 * Author: Bert Bos <bert@w3.org>
 * Created: 5 Nov 1998
 *
 * Input file for gperf, to generate a perfect hash function
 * for all HTML tags, and to store each element's type.
 *
 * mixed = element accepts text content
 * empty = element is empty
 * cdata = element has character data content (i.e., unparsed content)
 * stag = start tag is required
 * etag = end tag is required
 * pre = element is preformatted
 * break_before, break_after = pretty-print with a newline before/after the elt
 * parents = array of possible parents, first one is preferred parent
 *
 * The DTD is a mixture of strict HTML 4.0 and some HTML5
 *
 */
#include <config.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <stdbool.h>
#include "export.h"
#include "types.e"

#define MAXNAMELEN 10				/* Max. len. of elt. name */
EXPORTDEF(MAXNAMELEN)

EXPORT typedef struct _ElementType {
  string name;
  bool mixed, empty, cdata, stag, etag, pre, break_before, break_after;
  string parents[76];
} ElementType;

/* lookup_element -- look up the string in the hash table */
EXPORT const ElementType * lookup_element(/* register const char *str,
					  register unsigned int len */);

/* Different kinds of parent elements: */
#define PHRASE "abbr", "acronym", "b", "bdi", "bdo", "big", "cite", "code", "dfn", "em", "i", "kbd", "q", "s", "samp", "small", "span", "strong", "sub", "sup", "time", "tt", "u", "var"
#define BRIDGE "p", "address", "caption", "dt", "h1", "h2", "h3", "h4", "h5", "h6", "legend", "pre"
#define FLOW "article", "aside", "dd", "del", "details", "div", "fieldset", "figcaption", "figure", "footer", "header", "ins", "li", "nav", "section", "main", "object", "td", "th"
#define INLINE_PARENT BRIDGE, PHRASE, FLOW
#define BLOCK_PARENT "body", "blockquote", "map", "main"

#define TOTAL_KEYWORDS 97
#define MIN_WORD_LENGTH 1
#define MAX_WORD_LENGTH 10
#define MIN_HASH_VALUE 1
#define MAX_HASH_VALUE 227
/* maximum key range = 227, duplicates = 0 */

#ifdef __GNUC__
__inline
#else
#ifdef __cplusplus
inline
#endif
#endif
static unsigned int
hash (str, len)
     register const char *str;
     register unsigned int len;
{
  static const unsigned char asso_values[] =
    {
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228,   0, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228,  35,
       30,  20,  15,   5,   0, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228,   5,  60,  80,
        5,   0,  85,  37,  65,  20, 228,  15,  10,  10,
       25,  35,  25,   0,  80,  30,   0,  65, 105,   0,
      228,  40, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228, 228, 228, 228, 228,
      228, 228, 228, 228, 228, 228
    };
  register int hval = len;

  switch (hval)
    {
      default:
        hval += asso_values[(unsigned char)str[1]];
      /*FALLTHROUGH*/
      case 1:
        hval += asso_values[(unsigned char)str[0]];
        break;
    }
  return hval + asso_values[(unsigned char)str[len - 1]];
}

#ifdef __GNUC__
__inline
#if defined __GNUC_STDC_INLINE__ || defined __GNUC_GNU_INLINE__
__attribute__ ((__gnu_inline__))
#endif
#endif
const ElementType *
lookup_element (str, len)
     register const char *str;
     register unsigned int len;
{
  static const ElementType wordlist[] =
    {
      {""},
#line 130 "dtd.hash"
      {"q",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 153 "dtd.hash"
      {"tt",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""}, {""}, {""}, {""},
#line 90 "dtd.hash"
      {"dt",		1, 0, 0, 1, 0, 0, 1, 1, {"dl", NULL}},
      {""}, {""},
#line 143 "dtd.hash"
      {"table",		0, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
#line 62 "dtd.hash"
      {"a",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "label", NULL}},
#line 145 "dtd.hash"
      {"td",		1, 0, 0, 1, 0, 0, 1, 1, {"tr", NULL}},
#line 146 "dtd.hash"
      {"textarea",	1, 0, 0, 1, 1, 1, 0, 0, {INLINE_PARENT, "a", "label", NULL}},
      {""},
#line 61 "dtd.hash"
      {"%data",	1, 0, 0, 1, 0, 0, 0, 0, {"p", NULL}},
      {""},
#line 84 "dtd.hash"
      {"dd",		1, 0, 0, 1, 0, 0, 1, 1, {"dl", NULL}},
#line 85 "dtd.hash"
      {"del",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, BLOCK_PARENT, "a", "button", "noscript", "form", "label", "option", "textarea", NULL}},
#line 120 "dtd.hash"
      {"meta",		0, 1, 0, 1, 0, 0, 1, 0, {INLINE_PARENT, "head", "a", "button", "noscript", "label", NULL}},
#line 92 "dtd.hash"
      {"embed",		0, 1, 0, 1, 0, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 115 "dtd.hash"
      {"legend",		1, 0, 0, 1, 1, 0, 1, 1, {"fieldset", NULL}},
#line 91 "dtd.hash"
      {"em",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""}, {""},
#line 150 "dtd.hash"
      {"title",		1, 0, 0, 1, 1, 0, 1, 1, {"head", NULL}},
      {""},
#line 89 "dtd.hash"
      {"dl",		0, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
      {""}, {""},
#line 114 "dtd.hash"
      {"label",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", NULL}},
      {""}, {""}, {""}, {""},
#line 139 "dtd.hash"
      {"style",		1, 0, 1, 1, 1, 1, 1, 0, {"head", NULL}},
#line 134 "dtd.hash"
      {"select",		0, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "label", NULL}},
      {""}, {""}, {""},
#line 68 "dtd.hash"
      {"aside",		1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
#line 108 "dtd.hash"
      {"i",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 86 "dtd.hash"
      {"details",	1, 0, 0, 1, 1, 0, 1, 1, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 119 "dtd.hash"
      {"map",		0, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 118 "dtd.hash"
      {"main",		1, 0, 0, 1, 1, 0, 1, 1, {"body", "div", "noscript", NULL}},
#line 128 "dtd.hash"
      {"param",		0, 1, 0, 1, 0, 0, 1, 1, {"object", NULL}},
#line 113 "dtd.hash"
      {"keygen",		0, 1, 0, 1, 0, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 65 "dtd.hash"
      {"address",	1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
      {""},
#line 117 "dtd.hash"
      {"link",		0, 1, 0, 1, 0, 0, 1, 0, {"head", NULL}},
#line 110 "dtd.hash"
      {"input",		0, 1, 0, 1, 0, 0, 0, 0, {INLINE_PARENT, "a", "label", NULL}},
#line 127 "dtd.hash"
      {"p",		1, 0, 0, 1, 0, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
#line 116 "dtd.hash"
      {"li",		1, 0, 0, 1, 0, 0, 1, 1, {"ul", "ol", NULL}},
      {""}, {""},
#line 135 "dtd.hash"
      {"small",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""},
#line 124 "dtd.hash"
      {"ol",		0, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
      {""}, {""}, {""}, {""},
#line 133 "dtd.hash"
      {"section",	1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
      {""},
#line 131 "dtd.hash"
      {"samp",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""}, {""},
#line 103 "dtd.hash"
      {"h6",		1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
#line 122 "dtd.hash"
      {"noscript",	1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, BLOCK_PARENT, "a", "button", "noscript", "form", "label", "option", "textarea",  NULL}},
#line 71 "dtd.hash"
      {"base",		0, 1, 0, 1, 0, 0, 1, 1, {"head", NULL}},
#line 109 "dtd.hash"
      {"img",		0, 1, 0, 1, 0, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 136 "dtd.hash"
      {"source",		0, 1, 0, 1, 0, 0, 1, 1, {"audio", "video", "template", NULL}},
      {""},
#line 138 "dtd.hash"
      {"strong",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 104 "dtd.hash"
      {"head",		0, 0, 0, 0, 0, 0, 1, 1, {"html", NULL}},
#line 149 "dtd.hash"
      {"thead",		0, 0, 0, 1, 0, 0, 1, 1, {"table", NULL}},
      {""},
#line 102 "dtd.hash"
      {"h5",		1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
#line 111 "dtd.hash"
      {"ins",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, BLOCK_PARENT, "a", "button", "noscript", "form", "label", "option", "textarea",  NULL}},
#line 107 "dtd.hash"
      {"html",		0, 0, 0, 0, 0, 0, 1, 1, {NULL, NULL}},
#line 75 "dtd.hash"
      {"blockquote",	0, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
      {""}, {""},
#line 112 "dtd.hash"
      {"kbd",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 137 "dtd.hash"
      {"span",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""}, {""},
#line 154 "dtd.hash"
      {"ul",		0, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
#line 72 "dtd.hash"
      {"bdi",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""},
#line 147 "dtd.hash"
      {"tfoot",		0, 0, 0, 1, 0, 0, 1, 1, {"table", NULL}},
#line 126 "dtd.hash"
      {"option",		1, 0, 0, 1, 0, 0, 1, 1, {"select", "optgroup", NULL}},
#line 67 "dtd.hash"
      {"article",	1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
#line 125 "dtd.hash"
      {"optgroup",	0, 0, 0, 1, 1, 0, 1, 1, {"select", NULL}},
#line 66 "dtd.hash"
      {"area",		0, 1, 0, 1, 0, 0, 0, 0, {"map", NULL}},
      {""}, {""},
#line 101 "dtd.hash"
      {"h4",		1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
      {""}, {""},
#line 152 "dtd.hash"
      {"track",		0, 1, 0, 1, 0, 0, 1, 1, {"audio", "video", "template", NULL}},
#line 123 "dtd.hash"
      {"object",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "del", "header", "aside", "section", "main", "article", "nav", "fieldset", "head", "ins", "label", "object", NULL}},
#line 64 "dtd.hash"
      {"acronym",	1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 73 "dtd.hash"
      {"bdo",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 80 "dtd.hash"
      {"cite",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 144 "dtd.hash"
      {"tbody",		0, 0, 0, 0, 0, 0, 1, 1, {"table", NULL}},
      {""},
#line 100 "dtd.hash"
      {"h3",		1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
#line 129 "dtd.hash"
      {"pre",		1, 0, 0, 1, 1, 1, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
      {""},
#line 69 "dtd.hash"
      {"audio",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 95 "dtd.hash"
      {"figure",		1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
      {""},
#line 93 "dtd.hash"
      {"fieldset",	1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "form", NULL}},
      {""}, {""},
#line 132 "dtd.hash"
      {"script",		1, 0, 1, 1, 1, 1, 1, 0, {"body", INLINE_PARENT, "blockquote", "head", "map", "a", "button", "noscript", "form", "label", NULL}},
#line 79 "dtd.hash"
      {"caption",	1, 0, 0, 1, 1, 0, 1, 1, {"table", NULL}},
#line 87 "dtd.hash"
      {"dfn",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 81 "dtd.hash"
      {"code",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 74 "dtd.hash"
      {"big",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
#line 70 "dtd.hash"
      {"b",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""},
#line 142 "dtd.hash"
      {"sup",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""}, {""}, {""},
#line 99 "dtd.hash"
      {"h2",		1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
#line 82 "dtd.hash"
      {"col",		0, 1, 0, 1, 0, 0, 0, 0, {"colgroup", "table", NULL}},
      {""}, {""}, {""},
#line 148 "dtd.hash"
      {"th",		1, 0, 0, 1, 0, 0, 1, 1, {"tr", NULL}},
#line 88 "dtd.hash"
      {"div",		1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
#line 97 "dtd.hash"
      {"form",		0, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, NULL}},
      {""}, {""},
#line 98 "dtd.hash"
      {"h1",		1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
#line 121 "dtd.hash"
      {"nav",		1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
#line 76 "dtd.hash"
      {"body",		0, 0, 0, 0, 0, 0, 1, 1, {"html", NULL}},
#line 94 "dtd.hash"
      {"figcaption",	1, 0, 0, 1, 1, 0, 1, 1, {"figure", NULL}},
      {""},
#line 141 "dtd.hash"
      {"summary",	1, 0, 0, 1, 1, 0, 0, 0, {"details", NULL}},
#line 157 "dtd.hash"
      {"wbr",		0, 1, 0, 1, 0, 0, 0, 1, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""}, {""}, {""}, {""},
#line 83 "dtd.hash"
      {"colgroup",	0, 0, 0, 1, 1, 0, 1, 1, {"table", NULL}},
#line 63 "dtd.hash"
      {"abbr",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""},
#line 105 "dtd.hash"
      {"header",		1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
      {""}, {""}, {""}, {""},
#line 78 "dtd.hash"
      {"button",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "label", NULL}},
      {""},
#line 140 "dtd.hash"
      {"sub",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""}, {""}, {""},
#line 151 "dtd.hash"
      {"tr",		0, 0, 0, 1, 0, 0, 1, 1, {"tbody", "tfoot", "thead", NULL}},
      {""}, {""},
#line 156 "dtd.hash"
      {"video",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""},
      {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""},
      {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""},
#line 155 "dtd.hash"
      {"var",		1, 0, 0, 1, 1, 0, 0, 0, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""},
      {""}, {""}, {""},
#line 96 "dtd.hash"
      {"footer",		1, 0, 0, 1, 1, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}},
      {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""},
      {""}, {""}, {""}, {""}, {""}, {""},
#line 77 "dtd.hash"
      {"br",		0, 1, 0, 1, 0, 0, 0, 1, {INLINE_PARENT, "a", "button", "noscript", "label", NULL}},
      {""}, {""}, {""}, {""},
#line 106 "dtd.hash"
      {"hr",		0, 1, 0, 1, 0, 0, 1, 1, {BLOCK_PARENT, FLOW, "button", "noscript", "form", NULL}}
    };

  if (len <= MAX_WORD_LENGTH && len >= MIN_WORD_LENGTH)
    {
      register int key = hash (str, len);

      if (key <= MAX_HASH_VALUE && key >= 0)
        {
          register const char *s = wordlist[key].name;

          if (*str == *s && !strncmp (str + 1, s + 1, len - 1) && s[len] == '\0')
            return &wordlist[key];
        }
    }
  return 0;
}
#line 158 "dtd.hash"

